File: addon/src/material_types/diffuse_shader.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright © 2025 Surasia
import logging

from typing import cast
from bpy.types import (
    ShaderNodeGroup,
    ShaderNodeOutputMaterial,
    ShaderNodeTree,
)

from .material_types import MaterialType
from ..constants import EMPTY_CONTROL, EMPTY_TEXTURES
from ..json_definitions import CommonMaterial
from ..nodes.diffuse_shader import DiffuseShader
from ..utils import assign_value, create_image, create_node

__all__ = ["DiffuseShaderType"]


class DiffuseShaderType(MaterialType):
    def __init__(self, material: CommonMaterial, material_tree: ShaderNodeTree) -> None:
        self.material: CommonMaterial = material
        self.tree: ShaderNodeTree = material_tree
        self._create_nodes()

    def _get_textures(self, nodes: ShaderNodeGroup) -> None:
        if self.material["textures"].get("Color"):
            img = create_image(self.tree.nodes, 0, str(self.material["textures"]["Color"]))
            if img.image and img.image.colorspace_settings:
                img.image.colorspace_settings.name = "sRGB"  # pyright: ignore[reportAttributeAccessIssue]
            _ = self.tree.links.new(img.outputs[0], nodes.inputs[0])
            _ = self.tree.links.new(img.outputs[1], nodes.inputs[11])

        control = self.material["textures"].get("Control")
        if control and control != EMPTY_CONTROL:
            img = create_image(self.tree.nodes, -100, str(control))
            _ = self.tree.links.new(img.outputs[0], nodes.inputs[1])

        if self.material["textures"].get("Normal"):
            img = create_image(self.tree.nodes, -200, str(self.material["textures"]["Normal"]))
            _ = self.tree.links.new(img.outputs[0], nodes.inputs[2])

        alpha_map = self.material["textures"].get("AlphaMap")
        if alpha_map and alpha_map not in EMPTY_TEXTURES:
            img = create_image(self.tree.nodes, -300, str(alpha_map))
            _ = self.tree.links.new(img.outputs[0], nodes.inputs[11])

    def _create_nodes(self) -> None:
        shader = create_node(self.tree.nodes, 0, 0, ShaderNodeGroup)
        shader.node_tree = cast(ShaderNodeTree, DiffuseShader().node_tree)

        if not self.material["diffuse_info"]:
            logging.warning("Diffuse material has no diffuse_info key, this shouldn't happen!")
            return
        info = self.material["diffuse_info"]
        assign_value(shader, 3, info["roughness_white"])
        assign_value(shader, 4, info["roughness_black"])
        assign_value(shader, 5, info["metallic_white"])
        assign_value(shader, 6, info["metallic_black"])
        assign_value(shader, 7, (*info["si_color_tint"], 1.0))
        assign_value(shader, 8, info["si_amount"])
        assign_value(shader, 9, info["si_intensity"])
        assign_value(shader, 10, (*info["color_tint"], 1.0))

        self._get_textures(shader)
        material_output = create_node(self.tree.nodes, 200, 0, ShaderNodeOutputMaterial)
        _ = self.tree.links.new(shader.outputs[0], material_output.inputs[0])

File: ekur/src/materials/diffuse_shader.rs
/* SPDX-License-Identifier: GPL-3.0-or-later */
/* Copyright © 2025 Surasia */
use anyhow::Result;

use crate::definitions::material::MaterialTag;

use super::{
    common_utils::{f32_from_const, get_post_texture},
    serde_definitions::{DiffuseInfo, Material, ShaderType, TextureType},
};

pub(super) fn handle_diffuse_shader(mat: &MaterialTag, material: &mut Material) -> Result<()> {
    let post_process = mat.post_process_definition.elements.first();
    let mut diffuse_info = DiffuseInfo::default();
    if let Some(post_process) = post_process {
        material.shader_type = ShaderType::Diffuse;
        get_post_texture(post_process, material, 0, TextureType::Color)?;
        get_post_texture(post_process, material, 60, TextureType::Control)?;
        get_post_texture(post_process, material, 116, TextureType::Normal)?;
        get_post_texture(post_process, material, 196, TextureType::AlphaMap)?;
        diffuse_info.metallic_white = f32_from_const(material, 96)?;
        diffuse_info.metallic_black = f32_from_const(material, 100)?;
        diffuse_info.roughness_white = f32_from_const(material, 104)?;
        diffuse_info.roughness_black = f32_from_const(material, 108)?;
        material.diffuse_info = Some(diffuse_info);
    }
    material.shader_type = ShaderType::Diffuse;

    Ok(())
}

pub(super) fn handle_diffuse_decal_shader(
    mat: &MaterialTag,
    material: &mut Material,
) -> Result<()> {
    let post_process = mat.post_process_definition.elements.first();
    let mut diffuse_info = DiffuseInfo::default();
    if let Some(post_process) = post_process {
        material.shader_type = ShaderType::Diffuse;
        get_post_texture(post_process, material, 96, TextureType::Color)?;
        get_post_texture(post_process, material, 56, TextureType::Control)?;
        get_post_texture(post_process, material, 16, TextureType::Normal)?;
        diffuse_info.metallic_white = 0.5;
        diffuse_info.metallic_black = 1.0;
        diffuse_info.roughness_white = 0.5;
        diffuse_info.roughness_black = 1.0;
        diffuse_info.color_tint = (
            f32_from_const(material, 0)?,
            f32_from_const(material, 4)?,
            f32_from_const(material, 8)?,
        );
        material.diffuse_info = Some(diffuse_info);
    }
    material.shader_type = ShaderType::Diffuse;

    Ok(())
}

pub(super) fn handle_diffuse_decal_notint_shader(
    mat: &MaterialTag,
    material: &mut Material,
) -> Result<()> {
    let post_process = mat.post_process_definition.elements.first();
    let mut diffuse_info = DiffuseInfo::default();
    if let Some(post_process) = post_process {
        material.shader_type = ShaderType::Diffuse;
        get_post_texture(post_process, material, 48, TextureType::Color)?;
        get_post_texture(post_process, material, 8, TextureType::Control)?;
        diffuse_info.roughness_black = 0.25;
        diffuse_info.roughness_white = 0.20;
        diffuse_info.metallic_black = 0.25;
        diffuse_info.metallic_white = 0.20;
        material.diffuse_info = Some(diffuse_info);
    }
    material.shader_type = ShaderType::Diffuse;

    Ok(())
}

pub(super) fn handle_diffuse_billboard_shader(
    mat: &MaterialTag,
    material: &mut Material,
) -> Result<()> {
    let post_process = mat.post_process_definition.elements.first();
    let mut diffuse_info = DiffuseInfo::default();
    if let Some(post_process) = post_process {
        material.shader_type = ShaderType::Diffuse;
        get_post_texture(post_process, material, 96, TextureType::Color)?;
        get_post_texture(post_process, material, 160, TextureType::Control)?;
        get_post_texture(post_process, material, 128, TextureType::Normal)?;
        diffuse_info.metallic_white = f32_from_const(material, 76)?;
        diffuse_info.metallic_black = f32_from_const(material, 80)?;
        diffuse_info.roughness_white = f32_from_const(material, 60)?;
        diffuse_info.roughness_black = f32_from_const(material, 64)?;
        material.diffuse_info = Some(diffuse_info);
    }
    material.shader_type = ShaderType::Diffuse;

    Ok(())
}

pub(super) fn handle_diffuse_emissive_shader(
    mat: &MaterialTag,
    material: &mut Material,
) -> Result<()> {
    let post_process = mat.post_process_definition.elements.first();
    let mut diffuse_info = DiffuseInfo::default();
    if let Some(post_process) = post_process {
        material.shader_type = ShaderType::Diffuse;
        get_post_texture(post_process, material, 96, TextureType::Color)?;
        get_post_texture(post_process, material, 160, TextureType::Control)?;
        get_post_texture(post_process, material, 128, TextureType::Normal)?;
        get_post_texture(post_process, material, 244, TextureType::Emissive)?;
        diffuse_info.metallic_white = f32_from_const(material, 76)?;
        diffuse_info.metallic_black = f32_from_const(material, 80)?;
        diffuse_info.roughness_white = f32_from_const(material, 60)?;
        diffuse_info.roughness_black = f32_from_const(material, 64)?;
        diffuse_info.color_tint = (
            f32_from_const(material, 84)?,
            f32_from_const(material, 88)?,
            f32_from_const(material, 92)?,
        );
        diffuse_info.si_color_tint = (
            f32_from_const(material, 224)?,
            f32_from_const(material, 228)?,
            f32_from_const(material, 232)?,
        );
        diffuse_info.si_intensity = f32_from_const(material, 240)?;
        diffuse_info.si_amount = f32_from_const(material, 236)?;
        material.diffuse_info = Some(diffuse_info);
    }
    material.shader_type = ShaderType::Diffuse;

    Ok(())
}

pub(super) fn handle_diffuse_si_shader(mat: &MaterialTag, material: &mut Material) -> Result<()> {
    let post_process = mat.post_process_definition.elements.first();
    let mut diffuse_info = DiffuseInfo::default();
    if let Some(post_process) = post_process {
        material.shader_type = ShaderType::Diffuse;
        get_post_texture(post_process, material, 20, TextureType::Color)?;
        get_post_texture(post_process, material, 76, TextureType::Control)?;
        get_post_texture(post_process, material, 132, TextureType::Normal)?;
        diffuse_info.metallic_white = f32_from_const(material, 112)?;
        diffuse_info.metallic_black = f32_from_const(material, 116)?;
        diffuse_info.roughness_white = f32_from_const(material, 120)?;
        diffuse_info.roughness_black = f32_from_const(material, 124)?;
        diffuse_info.si_color_tint = (
            f32_from_const(material, 0)?,
            f32_from_const(material, 4)?,
            f32_from_const(material, 8)?,
        );
        diffuse_info.si_intensity = f32_from_const(material, 12)?;
        diffuse_info.si_amount = f32_from_const(material, 16)?;
        diffuse_info.color_tint = (
            f32_from_const(material, 48)?,
            f32_from_const(material, 52)?,
            f32_from_const(material, 56)?,
        );
        material.diffuse_info = Some(diffuse_info);
    }
    material.shader_type = ShaderType::Diffuse;
    Ok(())
}
